### PR TITLE
rpm: fix primary xml - part 2

### PR DESCRIPTION
**Pull request by urbanchef**
1. Add \<rpm:conflicts\> to primary.xml
2. [Skip primary package and provided files](https://github.com/rpm-software-management/createrepo_c/blob/master/src/parsehdr.c#L366) in \<rpm:requires\>

### CODE REVIEW

**Review by ylobankov — COMMENTED** [review]
The changes themselves look OK, but you need to provide some descriptions in the commit messages. Also, it would be nice to add some links to documentation or something else to understand why you do it.

File: rpmrepo.py [hunk 1]
```diff
@@ -877,6 +877,32 @@ def header_to_primary(
                                'ver': provides_ver,
                                'flags': provides_flags}
 
+    # files
```
Comment (ylobankov) [inline rpmrepo.py:880]: Please add a blank line after (for consistency with others).

File: rpmrepo.py [hunk 2]
```diff
@@ -877,6 +877,32 @@ def header_to_primary(
...
+    if not isinstance(basenames, list):
+        basenames = [basenames]
+    dirindexes = header.get('DIRINDEXES', [])
+    if not isinstance(dirindexes, list):
+        dirindexes = [dirindexes]
+    filemodes = header.get('FILEMODES', [])
+    if not isinstance(filemodes, list):
+        filemodes = [filemodes]
+
+    # represent integer as a C uint16 type value
```
Comment (ylobankov) [inline rpmrepo.py:894]: Please capitalize and add a dot to the end of the sentence.

File: rpmrepo.py [hunk 3]
```diff
@@ -902,12 +928,25 @@ def header_to_primary(
...
-                               'flags': requires_flags,
-                               "pre": pre}
+        nerv_entry = {'name': requires_name,
+                      'epoch': requires_epoch,
+                      'rel': requires_rel,
+                      'ver': requires_ver,
+                      'flags': requires_flags
+                      }
+
+        # Skip files which are provided
```
Comment (ylobankov) [inline rpmrepo.py:938]: Please add a dot to the end of the sentence.

File: rpmrepo.py [hunk 4]
```diff
@@ -902,12 +928,25 @@ def header_to_primary(
...
+        if nerv in provides_dict:
+            if nerv_entry == provides_dict[nerv]:
+                continue
+
+        # Skip package primary files
```
Comment (ylobankov) [inline rpmrepo.py:943]: Please add a dot to the end of the sentence.

File: rpmrepo.py [hunk 5]
```diff
@@ -902,12 +928,25 @@ def header_to_primary(
...
+        full_filenames = [f['name'] for f in files if f['type'] == 'file']
+        if requires_name in full_filenames and is_primary_file(requires_name):
+            continue
+
+        nerv_entry['pre'] = pre
```
Comment (ylobankov) [inline rpmrepo.py:948]: Why do you define `pre` here, but not above in `nerv_entry`?

File: rpmrepo.py [hunk 6]
```diff
@@ -547,6 +572,21 @@ def sort_key(item):
 
         res += '    </rpm:obsoletes>\n'
 
+        if fmt['conflicts']:
```
[the remark this answers is not in the document]
Comment (urbanchef) [reply rpmrepo.py:575]: provides and requires is not empty usually, whereas conflicts can often be. Additionally, createrepo_c does not include a tag if it does not contain entries. So I decided it should be conditional

File: rpmrepo.py (hunk 5 above)
Comment (urbanchef) [reply rpmrepo.py:948]: @ylobankov because there's dict comparison above at 940

`if nerv_entry == provides_dict[nerv]:`

`provides_dict[nerv]` does not have pre

Comment (ylobankov) [reply rpmrepo.py:949]: Ok, got it.

File: rpmrepo.py (hunk 1 above)
Comment (urbanchef) [reply rpmrepo.py:880]: Done

File: rpmrepo.py (hunk 2 above)
Comment (urbanchef) [reply rpmrepo.py]: Done

File: rpmrepo.py (hunk 3 above)
Comment (urbanchef) [reply rpmrepo.py]: Done

File: rpmrepo.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ylobankov) [reply rpmrepo.py:575]: Please share the link to the `createrepo_c` source code, where it doesn't include the tag if there are no entries.

[the remark this answers is not in the document]
Comment (urbanchef) [reply rpmrepo.py:575]: @ylobankov admittedly, createrepo_c is structured differently than mkrepo. [First](https://github.com/rpm-software-management/createrepo_c/blob/master/src/parsehdr.c#L337), on each iteration a corresponding RPMTAG_* header value is parsed. Second, createrepo_c creates a dynamic dependency object [here](https://github.com/rpm-software-management/createrepo_c/blob/master/src/parsehdr.c#L419). After that, if there's a DEP_CONFLICTS dependency type is met a data structure called `pkg->conflicts` is updated. 

Basically, if there's no DEP_CONFLICTS dependency, `pkg->conflicts` is never updated and it never makes it to primary.xml. This behaviour can be confirmed in practice.

[the remark this answers is not in the document]
Comment (ylobankov) [reply rpmrepo.py:575]: Ok, thanks.

**Review by ylobankov — APPROVED** [review]
LGTM
